filter.go

```diff
@@ -0,0 +1,280 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"slices"
+	"strings"
+	"unicode"
+)
+
+var (
+	ErrInternal         = errors.New("internal error")
+	ErrUnexpectedEOF    = errors.New("unexpected end of filter")
+	ErrUnexpectedInput  = errors.New("unexpected input")
+	ErrUnimplemented    = errors.New("unimplemented")
+	ErrInvalidAttrName  = errors.New("invalid attribute name")
+	ErrEmptyAttrName    = errors.New("empty attribute name")
+	ErrEmptyAttrValue   = errors.New("empty attribute value")
+	ErrMissingOperation = errors.New("missing filter operation")
+)
+
+// FilterNode represents an individual filter element of a parsed LDAP filter
+// string. A full filter is an abstract syntax tree (AST) made of FilterNodes.
+//
+// LDAP filters: https://ldap.com/ldap-filters/
+type FilterNode interface {
+	Match(e *Entry) bool
+}
+
+// Parse parses an LDAP filter string into an [FilterNode] AST that can be
+// used to match against an LDAP Entry. If the filter string could not be
+// parsed due to a syntax error or unimplemented filter, an error is
+// returned with a nil FilterNode.
+func Parse(filter string) (n FilterNode, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if perr, ok := r.(error); ok {
+				n, err = nil, perr
```

Comment:
what if there is a panic that does contain an error? (from within the library or a coding error) - shouldn't this have an else cause or panic again or something?

Reply:
i did it the opposite of "wrap a base ParseError". Instead the outer layer is a `parseError` which is wrapped by the functions that do the panic. And we unwrap if one of those and return what it wraps. Otherwise we just re-panic the recovered value.

Reply:
interesting, i think i like the outer layer parseError idea.